components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+    @Override
+    public PublisherDataHolder getSubscriber(String subscriberId, boolean returnSecrets) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement preparedStmt = new NamedPreparedStatement(connection, GET_SUBSCRIBER_SQL)) {
+                preparedStmt.setString(SUBSCRIBER_ID, subscriberId);
+                preparedStmt.setInt(TENANT_ID, tenantId);
+                try (ResultSet resultSet = preparedStmt.executeQuery()) {
+                    if (resultSet.next()) {
+                        return getPublisherDataHolder(resultSet, returnSecrets);
+                    }
+                    return null;
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException(String.format("Error while retrieving subscriber details of id : %s",
+                    subscriberId), e);
+        }
+    }
+
+    /**
+     * Gets all subscriber IDs.
+     *
+     * @param filter search string.
+     * @return list of subscriber IDs.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public List<String> listSubscriberIds(String filter) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+        List<String> subscriberIdList = new ArrayList<>();
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement preparedStmt = new NamedPreparedStatement(connection, GET_SUBSCRIBER_IDS_SQL)) {
+                preparedStmt.setInt(TENANT_ID, tenantId);
+
+                try (ResultSet subscriberIds = preparedStmt.executeQuery()) {
+                    while (subscriberIds.next()) {
+                        subscriberIdList.add(subscriberIds.getString(SUBSCRIBER_ID));
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while retrieving subscriber ids", e);
+        }
+
+        return EntitlementUtil.filterSubscribers(subscriberIdList, filter);
+    }
+
+    /**
+     * Adds a subscriber.
+     *
+     * @param holder publisher data holder.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public void addSubscriber(PublisherDataHolder holder) throws EntitlementException {
+
+        String subscriberId = EntitlementUtil.resolveSubscriberId(holder);
+        if (subscriberId == null) {
+            throw new EntitlementException(ERROR_SUBSCRIBER_ID_NULL);
+        }
+
+        if (isSubscriberExists(subscriberId)) {
+            throw new EntitlementException("Subscriber ID already exists");
+        }
+
+        PublisherPropertyDTO[] propertyDTOs = holder.getPropertyDTOs();
+        insertSubscriber(subscriberId, holder, propertyDTOs);
```

Comment:
Why we need to pass `propertyDTOs` separately?

Reply:
better if its one of following,
`insertSubscriber(subscriberId, holder);`
or
`insertSubscriber(subscriberId, moduleName, propertyDTOs);`